### PR TITLE
HLE: cellPngDec fixes, Common/DX12/Vulkan: texture formats improved

### PR DESCRIPTION
**Pull request by tambry**
cellPngDec commit implements support for converting RGBA to GA. This brings Closure to be on par with LLE cellPngDec.

Second commit drastically improves texture format support for Common and DX12. I'll also be porting these fixes to Vulkan (and maybe OGL).
Some screenshots for comparing the fixes (DX12):
PS3: http://i.imgur.com/NiwUfvm.png
DX12 Before: http://i.imgur.com/o8Kr2PS.png
DX12 After: http://i.imgur.com/sfXydrq.png
Vulkan Before: http://i.imgur.com/Ioa6u6C.png
Vulkan After: http://i.imgur.com/79OznwQ.png
~~OpenGL Before: http://i.imgur.com/oQYRiBG.png
OpenGL After: Soon...~~
Scratch that, OGL backend doesn't support proper remapping and stuff. It's very confusing compared to Vulkan and DX12. Not going to fix anything for it.

For DX12, there are some things I'm not sure how to fix: ie. 2nd row's 2nd (not sure what's wrong here, alpha portion seems fine) and 4th row's last (no native alternative format, will need to convert to some other format).

As for Vulkan, there are some non-native formats, that I'm unsure on how to handle properly. They'll require special conversion to other formats.
This though fixes text in Bit Trip Runner 2 and the logo for Closure.

Third commit is to fix the forced_unit printout being incorrect for unimplemented SCT/SCB instructions.

Also changed the log type for jstrchk to TODO, since it's not properly implemented.

### CODE REVIEW

**Comment by danilaml** [comment]
Two textures in the top right corner have R and B channels swapped. Is this easily fixable?

**Comment by tambry** [comment]
@vlj Any idea what I'm doing wrong with CELL_GCM_TEXTURE_A4R4G4B4 at https://github.com/RPCS3/rpcs3/pull/1653/files#diff-376ae528a7e121e12d628e35d470c603R146? Switching R and B screws up all colours.

**Comment by vlj** [comment]
Will have a look later.
Btw take attention that ARGB4 is a 16 bits formats.
ARGB8 is uploaded untouched to avoid the cost of swizzling but in your code ARGB4 is byte swapped (it's read as be_t<u16> and written as u16). So you likely need to use the swizzle format accordingly.

**Comment by tambry** [comment]
@vlj Thanks, managed to fix it now.

**Comment by tambry** [comment]
@vlj Could you review? Should be ready for merge.

File: rpcs3/Emu/RSX/Common/TextureUtils.cpp [hunk 1]
```diff
@@ -150,28 +150,44 @@ std::vector<rsx_subresource_layout> get_subresources_layout(const rsx::texture &
...
+	case CELL_GCM_TEXTURE_DEPTH24_D8_FLOAT: // Untested
+	case CELL_GCM_TEXTURE_D8R8G8B8:
+	case CELL_GCM_TEXTURE_A8R8G8B8:
+	case CELL_GCM_TEXTURE_Y16_X16:
+	case CELL_GCM_TEXTURE_Y16_X16_FLOAT:
+		return get_subresources_layout_impl<1, u32>(pixels, w, h, depth, layer, texture.get_exact_mipmap_count(), texture.pitch(), !is_swizzled);
+
 	case CELL_GCM_TEXTURE_W16_Z16_Y16_X16_FLOAT:
 		return get_subresources_layout_impl<1, u64>(pixels, w, h, depth, layer, texture.get_exact_mipmap_count(), texture.pitch(), !is_swizzled);
+
```
Comment (vlj) [inline rpcs3/Emu/RSX/Common/TextureUtils.cpp]: extra blank line

File: rpcs3/Emu/RSX/Common/TextureUtils.cpp [hunk 2]
```diff
@@ -150,28 +150,44 @@ std::vector<rsx_subresource_layout> get_subresources_layout(const rsx::texture &
...
 	case CELL_GCM_TEXTURE_COMPRESSED_DXT1:
 		return get_subresources_layout_impl<4, u64>(pixels, w, h, depth, layer, texture.get_exact_mipmap_count(), texture.pitch(), !is_swizzled);
+
```
Comment (vlj) [inline rpcs3/Emu/RSX/Common/TextureUtils.cpp]: extra blank line

File: rpcs3/Emu/RSX/Common/TextureUtils.cpp [hunk 3]
```diff
@@ -150,28 +150,44 @@ std::vector<rsx_subresource_layout> get_subresources_layout(const rsx::texture &
...
-		return get_subresources_layout_impl<1, u32>(pixels, w, h, depth, layer, texture.get_exact_mipmap_count(), texture.pitch(), !is_swizzled);
+	case CELL_GCM_TEXTURE_B8:
+		return get_subresources_layout_impl<1, u8>(pixels, w, h, depth, layer, texture.get_exact_mipmap_count(), texture.pitch(), !is_swizzled);
+
+	case ~(CELL_GCM_TEXTURE_LN | CELL_GCM_TEXTURE_UN) & CELL_GCM_TEXTURE_COMPRESSED_B8R8_G8R8:
+	case ~(CELL_GCM_TEXTURE_LN | CELL_GCM_TEXTURE_UN) & CELL_GCM_TEXTURE_COMPRESSED_R8B8_R8G8:
+	case CELL_GCM_TEXTURE_COMPRESSED_HILO8:
+	case CELL_GCM_TEXTURE_COMPRESSED_HILO_S8:
 	case CELL_GCM_TEXTURE_DEPTH16:
+	case CELL_GCM_TEXTURE_DEPTH16_FLOAT: // Untested
```
Comment (vlj) [inline rpcs3/Emu/RSX/Common/TextureUtils.cpp]: I would like to keep untested texture format support out of code. It's much harder to debug visual error when you don't know where to look for.

Comment (tambry) [inline rpcs3/Emu/RSX/Common/TextureUtils.cpp]: I only needed to add that, so the texture formats test would boot. It and other untested formats should work, but they should be verified. I could probably go through the trouble of updating the tests and the texture format generation script to also add support for generation of such formats. I'll look into it.

File: rpcs3/Emu/RSX/Common/TextureUtils.cpp (hunk 2 above)
Comment (tambry) [inline rpcs3/Emu/RSX/Common/TextureUtils.cpp]: Those are intentional. In my opinion they make the code more readable. Removed them though.

**Comment by vlj** [comment]
Need rebase then can be merged

**Comment by tambry** [comment]
@vlj I'm currently migrating the old Alex's texture formats test, which I used heavily for this PR, to a slightly nicer code base along with the scripts used for generating the textures. I'll probably be adding texture generation for the 4 missing DEPTH texture formats.
I'll be pushing the final version of the fixed up texture formats test to my [PS3Tests](https://github.com/tambry/PS3Tests) sometime tonight. (RSX/TextureFormats, TEST00008)

**Comment by tambry** [comment]
Sadly I was unable to figure out exactly how the DEPTH formats work. I've pushed the fixed tests to the repository, though. The current fixes should be good enough for merging, though.

**Comment by vlj** [comment]
depth is a rather opaque format, I don't think it's possible to upload a texture using this format. Or at least not easily.

**Pull request merged**